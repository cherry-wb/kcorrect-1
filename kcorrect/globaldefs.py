#!/usr/bin/env python2.6
"""
Global definitions
"""
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
#from __future__ import unicode_literals
import os
from numpy import float32


# make sure KCORRECT_DIR is set
KCORRECT_DIR = os.getenv('KCORRECT_DIR')
if KCORRECT_DIR is None:
    print('KCORRECT_DIR not defined.')
    exit()
elif not os.path.exists(KCORRECT_DIR):
    print('Problem accessing %s.' % KCORRECT_DIR)
    exit()


# default directories for some data sets
DEFAULT_FILTER_DIR = KCORRECT_DIR + '/data/filters'
DEFAULT_TEMPLATE_DIR = KCORRECT_DIR + '/data/templates'

# default cosmology (Omega_matter, Omega_lambda, h_100)
COSMO_DEFAULT = (0.3, 0.7, 0.7)

# default redshift range (zmin, zmax, # of points)
#ZRANGE_DEFAULT = (0.0, 2.0, 1000)
ZRANGE_DEFAULT = (0.0, 6.0, 3000)

# template name
VNAME = 'default'

# within the kcorrect C library, float C type is consistently used.
FTYPE = float32

# filter definition file extension
FILTEREXT = '.par'
