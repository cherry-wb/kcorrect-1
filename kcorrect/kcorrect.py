#!/usr/bin/env python2.6
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
#from __future__ import unicode_literals
import numpy as np
from kcorrect.photo import Photo
from kcorrect.utils.cosmology import ztodm
from kcorrect.utils.photo import maggie2ab, fit_nonneg, reconstruct_maggie
from kcorrect.globaldefs import FTYPE


__all__ = ['KCorrect', 'KCorrectAB']


class KCorrect(Photo):
    """
    k correction class for photometry in maggies

    This is the base class for k correction which implements the most
    basic functionalities.  AB maggies are handy to deal with
    internally, but they are not really the conventional photometric
    system observers are used to.  Therefore one usually wants to
    subclass KCorrect to implement a conversion from the input
    photometry to AB maggies by overriding set_data and convert2maggie
    methods.
    """

    def __init__(self, *args, **kwargs):
        """
        Parameters
        ----------
        ptable : ProjectionTableDB object
            Defaults to ProjectionTableDB().  If a non-standard
            projection table (i.e., with different redshift intervals
            or template spectra, etc.) is to be used in reconstructing
            AB maggies, then a custom ProjectionTableDB object can be
            specified here.
        cosmo : (Omega_matter, Omega_Lambda, h_100)
        maxiter : int
            Maximum number of iterations for non-negative fitting.
        verbosity : int
            Set the level of verbosity for output; currently not
            really used.
        """
        self.maxiter = kwargs['maxiter'] if 'maxiter' in kwargs else 3000
        super(KCorrect, self).__init__(*args, **kwargs)
        self._model_maggie = None
        self._distmod = None

    def set_input(self, redshift, filter_list, maggie, maggie_ivar,
                  *args, **kwargs):
        """
        Set input data

        INPUT

        redshift -- List of redshifts
        filter_list -- A FilterList object
        maggie -- List of AB maggies
        maggie_ivar -- List of inverse variances of AB maggies
        """
        super(KCorrect, self).set_input_photo(filter_list, maggie, maggie_ivar)

        redshift = np.atleast_1d(np.asarray(redshift, dtype=FTYPE))
        self.redshift = redshift.reshape((len(self), 1))

        # zero redshifts for convenience
        self.redshift0 = np.zeros(self.redshift.shape, dtype=FTYPE)

        # get fit coefficients once and for all
        self.do_fit(maxiter=self.maxiter)

    @property
    def model_maggie(self):
        """
        Maggies of the model SED through the input passbands.
        
        model_maggie only differs from input_maggie in that they are
        computed from the best fit SEDs through the input passbands;
        these two maggies should be very close to each other if the
        fits are good.
        """
        if self._model_maggie is not None:
            return self._model_maggie
        self._model_maggie = reconstruct_maggie(self.coeffs, self.redshift,
                                                self.ptable[self.filter_list])
        return self._model_maggie

    @property
    def distmod(self):
        """
        Distance moduli to input redshifts.
        """
        if self._distmod is not None:
            return self._distmod
        dm = ztodm(self.redshift, self.cosmo)
        self._distmod = np.asarray(dm, dtype=FTYPE)
        return self._distmod

    def do_fit(self, maxiter=3000):
        """
        Do non-negative fitting.
        """
        self._model_maggie = None
        r = fit_nonneg(self.input_maggie, self.input_maggie_ivar,
                       self.redshift, self.ptable[self.filter_list],
                       maxiter)
        self.coeffs, self.chi2, self.iter = r

    def kcorrect(self, filter_list_q=None, filter_list_r=None, band_shift=0.):
        """
        Compute the k-corrections between two sets of filters

        This method computes k corrections bewteen two sets of filters
        such that

        m_r = M_q + DM(z) + K_qr(z)

        where m_r is the apparent magnutude through filter r, M_q the
        absolute magnitude through filter q, DM the distance modulus,
        and K_qr is the k correction between filters q and r.

        The number of filters in sets q and r must match.

        INPUT

        filter_list_q, filter_list_r
            -- FilterList objects; if not provided explicitly, q and r
               both defaults to self.filter_list

        band_shift -- Amount of redshift used for computing the
                      absolute magnitudes through band-shifted
                      filters; defaults to zero.
        """
        if not filter_list_q: filter_list_q = self.filter_list
        if not filter_list_r: filter_list_r = self.filter_list

        if len(filter_list_q) != len(filter_list_r):
            raise ValueError('Numbers of filters must match.')

        if filter_list_r != self.filter_list:
            rm2 = reconstruct_maggie(self.coeffs, self.redshift,
                                     self.ptable[filter_list_r])
        else:
            rm2 = self.model_maggie

        if band_shift > 0.:
            bs = np.ones(self.redshift.shape, dtype=FTYPE) * band_shift
            rm1 = reconstruct_maggie(self.coeffs, bs,
                                     self.ptable[filter_list_q])
            rm1 = rm1 / (1. + bs)
        else:
            rm1 = reconstruct_maggie(self.coeffs, self.redshift0,
                                     self.ptable[filter_list_q])
        return 2.5 * np.log10(rm1 / rm2)

    def absmag(self, filter_list=None, band_shift=0.):
        """
        Compute the absolute AB magnitude through a set of filters

        This method returns the absolute AB magnitudes through a set
        of (band-shifted) filters, which defaults to the input
        filters.  The output magnitudes are computed from the input
        photometries where available; otherwise the output magnitudes
        are computed from the model photometries reconstructed from
        the best fit SED.

        INPUT

        filter_list -- A FilterList object; defaults to the input filters
        band_shift -- Amount of redshift used for computing the
                      absolute magnitudes through band-shifted
                      filters; defaults to zero.
        """
        if not filter_list:
            filter_list = self.filter_list

        # get model maggies for abs mag
        bs = np.ones(self.redshift.shape, dtype=FTYPE) * band_shift
        rm = reconstruct_maggie(self.coeffs, bs, self.ptable[filter_list])
        rm = rm / (1. + bs)

        # fill all entries with model
        mag = maggie2ab(rm) - self.distmod
        if len(filter_list) == len(self.filter_list):
            # now replace those entries with good input and well
            # defined ivar
            mask = (np.greater(self.input_maggie,0.) *
                    np.greater(self.input_maggie_ivar, 0.)).astype(int)
            kcorr = self.kcorrect(filter_list_q=filter_list,
                                  band_shift=band_shift)
            mag = ((1 - mask) * mag
                   + mask * (maggie2ab(self.input_maggie)
                             - self.distmod - kcorr))
        return mag

    def appmag(self, filter_list, redshift=[]):
        """
        Compute apparent AB magnitudes through a set of filters

        Computes the apparent magnitudes through a set of filters.
        Optionally, the objects to be observed can be moved to
        different redshifts than the input redshifts; i.e., the best
        fit SEDs are redshifted arbitrarily and *observed* with the
        set of filters.

        INPUT

        filter_list -- A FilterList object
        redshift -- A redshift or a 1D array of redshift (with the
                    number of elements equal to the number of input
                    objects) to observe galaxies at arbitrary
                    redshifts
        """
        redshift = np.atleast_1d(redshift).astype(FTYPE)
        if redshift.size == 0:
            # simply observe the input objects with given filters
            maggie = reconstruct_maggie(self.coeffs, self.redshift,
                                        self.ptable[filter_list])
            return maggie2ab(maggie)

        # when redshifts are given, use them
        if redshift.size == 1:
            redshift = np.ones(self.redshift.shape, dtype=FTYPE) * redshift[0]
        elif redshift.size == self.redshift.size:
            redshift = np.reshape(redshift, self.redshift.shape)
        else:
            raise ValueError('Inconsistent redshift array')
        maggie_S = reconstruct_maggie(self.coeffs, redshift,
                                      self.ptable[filter_list])
        m_S = maggie2ab(maggie_S)
        dm1 = ztodm(self.redshift, self.cosmo)
        dm2 = ztodm(redshift, self.cosmo)

        # because of what reconstruct_maggie returns, the flux needs
        # to be rescaled by the ratio of distances
        return m_S - dm1 + dm2

    def model_spectrum(self, index=None, lmin=None, lmax=None):
        """
        Returns the best fit SED(s) in the rest frame.

        The output SED is in units of erg s^-1 cm^-2 A^-1.

        Parameters
        ----------
        index : array of indices
            The index(ices) of the object for which spectrum is obtained.
        lmin : float
            Minimum wavelength.
        lmax : float
            Maximum wavelength.
        """
        index = (np.arange(0, len(self)) if index is None
                 else np.atleast_1d(np.asarray(index, int)))

        lamb, vmatrix = self.ptable[self.filter_list].templates
        lamb = 0.5 * (lamb[:-1] + lamb[1:])

        mask = np.ones(lamb.shape, dtype=bool)
        if lmin is not None:
            mask = mask * np.greater(lamb, lmin)
        if lmax is not None:
            mask = mask * np.less(lamb, lmax)
        lamb = lamb[mask]

        vmatrix = np.compress(mask, np.transpose(vmatrix), axis=0)

        data = []
        for coeffs in self.coeffs[index]:
            flux = np.add.reduce(np.transpose(vmatrix * coeffs))
            data.append(flux)
        return (lamb, data if len(data) > 1 else data[0])


class KCorrectAB(KCorrect):
    """
    K-correction for AB photometry

    Expect AB magnitudes and 1-sigma uncertainties
    """

    def set_input(self, filter_list, abmag, abmag_sigma, redshift,
                  *args, **kwargs):
        """
        Set input data

        If abmag_sigma is not given, the minimum uncertainties of 0.02
        mag will be assumed.

        INPUT

        redshift -- List of redshifts
        filter_list -- FilterList object
        abmag -- List of AB magnitudes
        abmag_sigma -- List of 1-sigma uncertainties in AB magnitudes;
                       defaults to an empty list
        """

        # if uncertainties not given, assume 2% errors
        abmag_sigma = (np.ones(abmag.shape, dtype=FTYPE) * 0.02
                       if abmag_sigma is None else abmag_sigma)

        # KCorrect.set_data expects maggies
        maggie, maggie_sigma = ab2maggie(abmag, abmag_sigma)
        maggie, maggie_ivar = maggie, sigma2ivar(maggie_sigma)

        super(KCorrect, self).set_input(filter_list, maggie, maggie_ivar,
                                        redshift, *args, **kwargs)
