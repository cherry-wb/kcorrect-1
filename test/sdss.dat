# redshift redshift_err u g r i z du dg dr di dz eu eg er ei
9.900e-02 1.290e-04 20.038 18.450 17.581 17.267 16.833 0.231 0.030 0.018 0.022 0.053 0.103 0.076 0.055 0.042 0.030
1.440e-01 1.379e-04 20.199 18.652 17.619 17.173 16.878 0.187 0.022 0.015 0.016 0.035 0.101 0.075 0.054 0.041 0.029
1.230e-01 1.464e-04 20.611 18.817 17.812 17.375 17.018 0.186 0.017 0.012 0.012 0.028 0.237 0.175 0.127 0.096 0.068
9.700e-02 1.418e-04 19.938 18.082 17.154 16.643 16.404 0.218 0.027 0.014 0.018 0.039 0.159 0.117 0.085 0.064 0.046
1.500e-01 1.587e-04 20.487 18.815 17.668 17.222 16.840 0.253 0.024 0.017 0.017 0.044 0.082 0.060 0.044 0.033 0.023
8.800e-02 1.785e-04 18.672 17.519 16.955 16.613 16.607 0.113 0.016 0.012 0.016 0.063 0.168 0.124 0.090 0.068 0.048
1.290e-01 1.552e-04 18.567 17.852 17.055 16.677 16.487 0.110 0.023 0.016 0.021 0.061 0.128 0.094 0.068 0.052 0.037
8.500e-02 5.798e-05 19.179 17.752 17.022 16.597 16.338 0.108 0.011 0.008 0.010 0.030 0.337 0.248 0.180 0.137 0.097
1.360e-01 1.556e-04 19.250 17.870 17.093 16.735 16.452 0.122 0.017 0.010 0.013 0.037 0.044 0.032 0.023 0.018 0.013
1.340e-01 1.674e-04 20.167 18.251 17.175 16.721 16.395 0.270 0.022 0.013 0.013 0.033 0.286 0.211 0.153 0.116 0.082
