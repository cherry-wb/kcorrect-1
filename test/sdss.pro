openr, lun, 'sdss.dat', /GET_LUN
header = STRARR(1)
READF, lun, header

rows = 10
data = FLTARR(17, rows)  
READF, lun, data

redshift = data(0,*)
lup = data(2:6,*)
lup_sig = data(7:11,*)
red = data(12:16,*)


lupcorr = lup - red

mag = fltarr(5,10)
mag_ivar = fltarr(5,10)
k_sdssfix,lupcorr,lup_sig,mag,mag_ivar

nmag = mag * 1e9
nmag_ivar = mag_ivar / 1e9 / 1e9

kcorr = fltarr(5,10)
kcorr = sdss_kcorrect(redshift,nmgy=nmag,ivar=nmag_ivar,absmag=absmag)

print, '### k-corrections: SDSS ugriz ###'
print, kcorr
print, ''

print, '### absolute magnitudes: SDSS ugriz ###'
print, absmag 
print, ''

kcorr = sdss_kcorrect(redshift,nmgy=nmag,ivar=nmag_ivar,absmag=absmag, $
                      coeffs=coeffs, band_shift=0.1, $
                      omaggies=om,oivar=oi)
print, '### k-corrections: SDSS ugriz band-shifted to z=0.1 ###'
print, kcorr
print, ''

print, '### absolute magnitudes: SDSS ugriz band-shifted to z=0.1 ###'
print, absmag
print, ''

; need to get vmatrix, lambda, and coeffs
kcorrect, om,oi, redshift, kcdm, vmatrix=vmatrix, lambda=lambda, $
  coeffs=coeffs, rmatrix=rmatrix, zvals=zvals

ngals = n_elements(redshift)

redshift0 = replicate(0.,n_elements(redshift))
distmod = lf_distmod(redshift)
distmod0 = lf_distmod(redshift0)

fl_sdss = 'sdss_'+['u','g','r','i','z']+'0.par'
fl = 'bessell_'+['U','B','V','R','I']+'.par'

k_reconstruct_maggies, coeffs, redshift0, rm0, $
  vmatrix=vmatrix, lambda=lambda, filterlist=fl
k_reconstruct_maggies, coeffs, redshift, rm1, $
  vmatrix=vmatrix, lambda=lambda, filterlist=fl_sdss

kcorr = 2.5*alog10(rm0/rm1)

abmag = fltarr(n_elements(fl),ngals)

for i = 0L, n_elements(fl)-1L do $
    abmag[i,*] = -2.5*alog10(mag[i,*]) - distmod - kcorr[i,*]

print, '### absolute magnitudes: Bessell UBVRI ###'
print, abmag
print, ''

band_shift = 0.1
bs = replicate(band_shift,ngals)
k_reconstruct_maggies, coeffs, bs, rm0, $
  vmatrix=vmatrix, lambda=lambda, filterlist=fl ;, band_shift=band_shift
rm0 = rm0 / (1.+band_shift)
kcorr = 2.5*alog10(rm0/rm1)
for i = 0L, n_elements(fl)-1L do $
    abmag[i,*] = -2.5*alog10(mag[i,*]) - distmod - kcorr[i,*]

print, '### absolute magnitudes: Bessell UBVRI band-shifted to z=0.1 ###'
print, abmag
print, ''

fl_deep = 'deep_'+['B','R','I']+'.par'
k_reconstruct_maggies, coeffs, redshift, maggies, $
  vmatrix=vmatrix, lambda=lambda, filterlist=fl_deep
appmag = -2.5*alog10(maggies)
print, '### apparent magnitudes: DEEP BRI ###'
print, appmag
print, ''

vega2ab = k_vega2ab(filterlist=fl_deep,/kurucz)
vegamag = fltarr(n_elements(fl_deep), ngals)
for i = 0L, n_elements(fl_deep)-1L do $
  vegamag[i,*] = appmag[i,*] - vega2ab[i]
print, '### apparent magnitudes: DEEP BRI on Vega scale ###'
print, vegamag
print, ''

redshift1 = replicate(1.0, n_elements(redshift))
distmod1 = lf_distmod(redshift1)
k_reconstruct_maggies, coeffs, redshift1, maggies, $
  vmatrix=vmatrix, lambda=lambda, filterlist=fl_deep
for i = 0L, n_elements(fl_deep)-1L do $
  appmag[i,*] = -2.5*alog10(maggies[i,*]) - distmod + distmod1
print, '### apparent magnitudes: DEEP BRI if at z = 1 ###'
print, appmag
print, ''

exit
