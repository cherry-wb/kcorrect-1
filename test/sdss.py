#!/usr/bin/env python2.6
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
#from __future__ import unicode_literals
import numpy as np
import matplotlib.pyplot as plt
import kcorrect as KC
from kcorrect.sdss import SDSSKCorrect, SDSSPhotoZ


def main():
    d = np.loadtxt('test/sdss.dat', dtype=float)
    z = d[:, 0]
    lup = d[:, 2:7]
    dlup = d[:, 7:12]
    red = d[:, 12:17]

    ####
    kc = SDSSPhotoZ(lup, dlup, red)
    print(kc.redshift)
    print(z)
    plt.plot(z, kc.redshift, 'o')
    plt.show()
    ####

    kc = SDSSKCorrect(z, lup, dlup, red, cosmo=(0.3, 0.7, 1.0))

    print('### k-corrections: SDSS ugriz ###')
    kcorr = kc.kcorrect()
    for each in kcorr: print('%+.6f %+.6f %+.6f %+.6f %+.6f' % tuple(each))
    print()

    print('### absolute magnitudes: SDSS ugriz ###')
    absmag = kc.absmag()
    for each in absmag: print('%+.4f %+.4f %+.4f %+.4f %+.4f' % tuple(each))
    print()

    print('### k-corrections: SDSS ugriz band-shifted to z=0.1 ###')
    band_shift = 0.1
    kcorr = kc.kcorrect(band_shift=band_shift)
    for each in kcorr: print('%+.6f %+.6f %+.6f %+.6f %+.6f' % tuple(each))
    print()

    print('### absolute magnitudes: SDSS ugriz band-shifted to z=0.1 ###')
    absmag = kc.absmag(band_shift=band_shift)
    for each in absmag: print('%+.4f %+.4f %+.4f %+.4f %+.4f' % tuple(each))
    print()

    print('### absolute magnitudes: Bessell UBVRI ###')
    fl = KC.FilterList(['bessell_U','bessell_B','bessell_V',
                        'bessell_R','bessell_I'])
    absmag = kc.absmag(fl)
    for each in absmag: print('%+.4f %+.4f %+.4f %+.4f %+.4f' % tuple(each))
    print()

    print('### absolute magnitudes: Bessell UBVRI band-shifted to z=0.1 ###')
    absmag = kc.absmag(fl,band_shift)
    for each in absmag: print('%+.4f %+.4f %+.4f %+.4f %+.4f' % tuple(each))
    print()

    print('### apparent magnitudes: DEEP BRI ###')
    fl = KC.FilterList(['deep_B','deep_R','deep_I'])
    fl.plot()
    mag = kc.appmag(fl)
    for each in mag: print('%+.4f %+.4f %+.4f' % tuple(each))
    print()

    print('### apparent magnitudes: DEEP BRI on Vega scale ###')
    vega2ab = KC.vega2ab(fl)
    mag = kc.appmag(fl) - vega2ab
    for each in mag: print('%+.4f %+.4f %+.4f' % tuple(each))
    print()

    print('### apparent magnitudes: DEEP BRI if at z = 1 ###')
    mag = kc.appmag(fl,redshift=1.0)
    for each in mag: print('%+.4f %+.4f %+.4f' % tuple(each))
    print()


if __name__ == '__main__':
    main()
