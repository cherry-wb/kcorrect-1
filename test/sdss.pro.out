redefining CTRL-D!!!!
redefining CTRL-K!!!!
  K_PROJECTION_TABLE: Creating rmatrix ...
  K_PROJECTION_TABLE: Done.
MRDFITS: Image array (5)  Type=Real*4
MRDFITS: Image array (5)  Type=Real*4
MRDFITS: Image array (5)  Type=Real*4
MRDFITS: Image array (5)  Type=Real*4
MRDFITS: Image array (5)  Type=Real*4
   K_READ_BASEL: 1 block(s) of spectra
    K_PROJECTION_TABLE: Creating rmatrix ...
    K_PROJECTION_TABLE: Done.
### k-corrections: SDSS ugriz ###
     0.238001     0.246573    0.0868333    0.0537998    0.0206644
     0.448305     0.446345     0.163623    0.0800017    0.0477281
     0.422750     0.373765     0.140260    0.0678796    0.0376901
     0.349963     0.271423     0.129958    0.0636063    0.0557124
     0.593746     0.521070     0.191316     0.112123    0.0695031
     0.160249     0.122534    0.0765692   -0.0852810   -0.0124663
    0.0705371     0.254232     0.120769    0.0510130    0.0252052
     0.190085     0.155164    0.0848075  -0.00452997    0.0287303
     0.303930     0.303014     0.121744    0.0233393    0.0182140
     0.501029     0.431245     0.153409    0.0811040    0.0411358

### absolute magnitudes: SDSS ugriz ###
     -17.8562     -19.3780     -20.0682     -20.3182     -20.6951
     -18.7807     -20.2520     -20.9833     -21.3146     -21.5533
     -18.1075     -19.7429     -20.4684     -20.7840     -21.0708
     -18.0770     -19.7646     -20.5211     -20.9268     -21.1279
     -18.7158     -20.2455     -21.0488     -21.3866     -21.7039
     -18.9377     -19.9610     -20.4471     -20.5872     -20.6340
     -19.8025     -20.6192     -21.2588     -21.5330     -21.6702
     -18.5498     -19.8049     -20.3985     -20.6732     -20.9134
     -19.3934     -20.7125     -21.3012     -21.5378     -21.7987
     -18.8804     -20.6038     -21.3460     -21.6727     -21.9127

  K_PROJECTION_TABLE: Creating rmatrix ...
  K_PROJECTION_TABLE: Done.
MRDFITS: Image array (5)  Type=Real*4
MRDFITS: Image array (5)  Type=Real*4
MRDFITS: Image array (5)  Type=Real*4
MRDFITS: Image array (5)  Type=Real*4
MRDFITS: Image array (5)  Type=Real*4
   K_READ_BASEL: 1 block(s) of spectra
    K_PROJECTION_TABLE: Creating rmatrix ...
    K_PROJECTION_TABLE: Done.
### k-corrections: SDSS ugriz band-shifted to z=0.1 ###
    -0.106191    -0.106687    -0.105546    -0.104767    -0.103897
    0.0502219    0.0623580   -0.0482323   -0.0795294   -0.0907905
   -0.0157934   -0.0162084   -0.0719880   -0.0916021    -0.100358
    -0.114757    -0.120002    -0.108146    -0.103479    -0.109119
     0.136906     0.104537   -0.0297006   -0.0704698   -0.0810336
    -0.116111    -0.127032    -0.106533    -0.114347    -0.100545
   -0.0906515   -0.0339719   -0.0732857   -0.0770901    -0.100614
    -0.127743    -0.146408    -0.118340    -0.120626    -0.102926
   -0.0297022  -0.00437116   -0.0734771   -0.0852673    -0.100407
    0.0449000    0.0321216   -0.0582111   -0.0801153   -0.0952466

### absolute magnitudes: SDSS ugriz band-shifted to z=0.1 ###
     -17.5120     -19.0247     -19.8759     -20.1596     -20.5705
     -18.3826     -19.8680     -20.7714     -21.1551     -21.4148
     -17.6690     -19.3530     -20.2562     -20.6246     -20.9328
     -17.6123     -19.3732     -20.2830     -20.7597     -20.9630
     -18.2589     -19.8290     -20.8278     -21.2040     -21.5534
     -18.6614     -19.7115     -20.2640     -20.5581     -20.5459
     -19.6414     -20.3310     -21.0647     -21.4049     -21.5444
     -18.2319     -19.5033     -20.1954     -20.5571     -20.7818
     -19.0598     -20.4051     -21.1060     -21.4292     -21.6801
     -18.4243     -20.2047     -21.1343     -21.5114     -21.7763

 K_PROJECTION_TABLE: Creating rmatrix ...
 K_PROJECTION_TABLE: Done.
MRDFITS: Image array (5)  Type=Real*4
MRDFITS: Image array (5)  Type=Real*4
MRDFITS: Image array (5)  Type=Real*4
MRDFITS: Image array (5)  Type=Real*4
MRDFITS: Image array (5)  Type=Real*4
  K_READ_BASEL: 1 block(s) of spectra
   K_PROJECTION_TABLE: Creating rmatrix ...
   K_PROJECTION_TABLE: Done.
 K_PROJECTION_TABLE: Creating rmatrix ...
 K_PROJECTION_TABLE: Done.
 K_PROJECTION_TABLE: Creating rmatrix ...
 K_PROJECTION_TABLE: Done.
### absolute magnitudes: Bessell UBVRI ###
     -17.9108     -19.1082     -19.8106     -20.0844     -20.5518
     -18.8423     -19.9545     -20.7026     -21.0697     -21.3941
     -18.1735     -19.4416     -20.1872     -20.5393     -20.9122
     -18.1572     -19.4625     -20.2110     -20.6593     -20.9386
     -18.7754     -19.9215     -20.7529     -21.1182     -21.5310
     -19.0117     -19.7773     -20.2247     -20.4615     -20.5321
     -19.8395     -20.3913     -21.0058     -21.3202     -21.5253
     -18.6210     -19.5790     -20.1394     -20.4644     -20.7620
     -19.4646     -20.4818     -21.0512     -21.3412     -21.6622
     -18.9453     -20.2945     -21.0646     -21.4274     -21.7559

 K_PROJECTION_TABLE: Creating rmatrix ...
 K_PROJECTION_TABLE: Done.
### absolute magnitudes: Bessell UBVRI band-shifted to z=0.1 ###
     -17.5395     -18.6258     -19.5942     -19.9059     -20.4046
     -18.4162     -19.4373     -20.4759     -20.8801     -21.2382
     -17.7064     -18.9097     -19.9574     -20.3494     -20.7565
     -17.6513     -18.9353     -19.9729     -20.4526     -20.7659
     -18.2995     -19.3548     -20.5019     -20.9129     -21.3595
     -18.6802     -19.4326     -20.1091     -20.3433     -20.4586
     -19.6540     -20.0512     -20.8435     -21.1545     -21.3895
     -18.2550     -19.1681     -19.9612     -20.2994     -20.6293
     -19.0844     -20.0598     -20.8749     -21.1835     -21.5389
     -18.4640     -19.7489     -20.8315     -21.2370     -21.6002

 K_PROJECTION_TABLE: Creating rmatrix ...
 K_PROJECTION_TABLE: Done.
### apparent magnitudes: DEEP BRI ###
      18.7769      17.4171      17.0696
      18.9751      17.4161      17.0208
      19.0546      17.5371      17.1546
      18.3756      16.9047      16.4673
      19.2621      17.4632      17.0390
      17.6029      16.7316      16.5704
      18.0126      16.8774      16.5250
      17.7667      16.6793      16.3877
      18.1314      16.9557      16.6183
      18.4617      16.8714      16.4861

 K_READ_BASEL: 1 block(s) of spectra
  K_PROJECTION_TABLE: Creating rmatrix ...
  K_PROJECTION_TABLE: Done.
### apparent magnitudes: DEEP BRI on Vega scale ###
      18.8895      17.1766      16.6118
      19.0877      17.1756      16.5629
      19.1672      17.2966      16.6968
      18.4882      16.6642      16.0095
      19.3747      17.2227      16.5811
      17.7156      16.4911      16.1126
      18.1252      16.6369      16.0672
      17.8794      16.4388      15.9299
      18.2441      16.7152      16.1605
      18.5743      16.6309      16.0283

 K_PROJECTION_TABLE: Creating rmatrix ...
 K_PROJECTION_TABLE: Done.
### apparent magnitudes: DEEP BRI if at z = 1 ###
      26.7772      24.8850      23.7996
      25.8787      24.1633      22.9705
      27.3529      24.7714      23.5014
      27.2428      24.7972      23.4690
      27.3943      24.3474      23.0337
      24.7591      23.9257      23.0191
      23.2545      23.0277      22.4085
      25.5878      24.2962      23.2754
      24.8748      23.4338      22.3828
      26.8808      23.9537      22.6526

